wallet_migrator/src/lib.rs

```diff
@@ -0,0 +1,65 @@
+pub mod credx2anoncreds;
+pub mod error;
+pub mod vdrtools2credx;
+
+use std::fmt::Display;
+
+use error::MigrationResult;
+use log::{debug, error, info, warn};
+pub use vdrtools::{
+    types::domain::wallet::{Config, Credentials, KeyDerivationMethod, Record},
+    Locator, WalletHandle,
+};
+
+/// Retrieves all records from a wallet and migrates them
+/// by applying the `migrate_fn` argument.
+///
+/// The migrated records are inserted into a newly created
+/// wallet, based on the provided `config` and `credentials`.
+pub async fn migrate_wallet<E>(
+    wallet_handle: WalletHandle,
+    config: Config,
+    credentials: Credentials,
+    migrate_fn: impl FnMut(Record) -> Result<Option<Record>, E>,
+) -> MigrationResult<()>
+where
+    E: Display,
+{
+    info!("Starting wallet migration...");
+    let locator = Locator::instance();
+
+    debug!("Creating new wallet {} ...", config.id);
+    locator
```

Comment:
Can do. I just thought that taking care of the wallet creation would mean we *always* create a new one and *ensure* we don't mess up the old one.

How is the users thing a problem, though? You have to pass the `Config` and the `Credentials` to construct the new wallet, which can be virtually anything.

Reply:
Touche, you are right, you can provide `Config` and `Credentials` with different db, different users etc. 

Thinking about how this migration API will be propagate to `vcx-napi-rs`, and how you would work with that from a NodeJS code in practice, how would you execute migration in server environment, I would find it cleaner if we take 
```
    source_config: Config,
    source_credentials: Credentials,
    target_config: Config,
    target_credentials: Credentials,
```
as  arguments into this function, and as suggested before, in the beginning of the function check that both wallets exists and be connected to, then proceed with migration, do not perform deletion at the end of the process.

That would give caller more flexibility how to approach creation of new wallet, how to possibly do dry runs with sqlite wallets, make the process for caller bit safer and "phased" by enabling him to create wallet himself first.

Reply:
I don't see how that is better in any way, tbh. If they provide the wallet handle for the source wallet, that means they already created and maybe populated one.

If we get the arguments to open a wallet, we might have to treat the fact that the wallet might not even exist.

Reply:
Imagine the code calling this, it will be working with the exact same arguments as I flashed out above, but then for the source wallet open it first, deal with wallet handle, but not for the target one. Why such a assymmetric approach?

You could say that "because the source wallet already exists, and the target one doesn't" - but I argue against that as well, and target wallet should also already exist. Migrator should just migrate IMO, not do other things - like create or try to delete a wallet, for the reasons argued above - perhaps I first need to convince you about this.

Reply:
My reasons for finding this the way to go are:

1) As you said, the source wallet must already exist. There's no point in being provided with a set of `config` and `credentials` to open the wallet and maybe even have to deal with it not existing at all. The purpose of the migrator is to migrate existent data, and for that a `WalletHandle` is sufficient. Creating a source wallet is in no way the responsibility of the migrator.

2) The destination wallet is a different story. I find this whole thing to be sensitive enough that we cannot afford data loss or inconsistencies, and no matter how much we test this we should always warn about that risk (which is what ACA-Py also does).  

    So, if we accept a `WalletHandle` there's nothing preventing a user from messing up and providing the same handle as for the source wallet. We could add a simple equality check, but that's not enough either. A wallet can be opened multiple times and you'd get different handles pointing to the same underlying wallet. 

    In contrast, if we ask for a set of `config` and `credentials` which we can use to create a new wallet, there's 0% chance of migrating within the source wallet, because the creation would fail if the same wallet info is provided (honestly I haven't directly tested this, but given that a new schema is created for the wallet I am confident that the database would throw an error on trying to create a duplicate). 

    So we are absolutely sure that whatever happens, no matter the error, only the destination wallet can get messed up and the source wallet will remain intact. We do our best to cleanup the destination wallet if that happens, but even if we fail to delete it, it's not the end of the world and the users can clean up. As for the source wallet, in the event of a successful migration consumers can delete it themselves (preferably after thorough testing).

3) You could also argue that the migrator does not delete records, so even if the same wallet is used as source and destination and an error occurs somewhere, the old records would still be there (this assuming that we use only `WalletHandle`s and remove the wallet deletion code on failure). And you'd be right. But the records that *did* get migrated will cause duplicate errors to occur on subsequent migrations. We can definitely handle all this, but this adds overhead, increases complexity, and all for, in my view, no good reason.

I personally think this more than justifies the asymmetric function arguments, which I honestly find to be quite a small thing anyway. I understand that maybe modifying this might make it easier for testing purposes and whatnot, but I'll always stand by my belief that the tests should accommodate production code, not the other way around.

Don't take this the wrong way, but I really doubt that my opinion on this will change. I am waaaaaaay more concerned with what calling this does than what the code that calls this looks like. So maybe we can loop @mirgee and @gmulhearn  to chip in and whatever the majority decides, we implement.

Reply:
Just adding the update of our discussion here:

We're still not sure how to parametrize this function, although I guess we'd all rather have two `Config` and `Credentials` sets, one for the old wallet and one for the new wallet, rather than two `WalletHandle`s. The only downside is that we have to error out if the source wallet does not exist (which is by no means a big deal, we just try to open it).

If people would rather see it like that, then I'll implement it that way.

Reply:
So we had some internal discussion and I am open to leaving things as they are. 

Some of my concern was also coming from the fact that create wallet (for mysql wallet type) now also create db schemas. But it does make sense considering sqlite type performs analogical setup, and if we ever need to migrate the schema, we can support that directly from the vdrtools wallet code.

Reply:
So to summarise, my very limited understanding of the tradeoffs in consideration here are: either

1. both the source and the target wallets are created and opened outside of the migration script and two wallet handles are passed in, or
2. both the source and the target wallets are created and opened outside of the migration script and two sets of configs are passed in, or
3. the source wallet is opened outside, destination wallet created and opened inside the script and source wallet handle and dest. wallet config are passed in.

As for 1., on one hand it makes sense to decouple opening the source wallet from the actual migration, but on the other hand it is done at the cost of exposing a lower-level primitive - the wallet handle.

As for 2., it shields the user away from the wallet handles, mitigates the danger of source and wallet handle being of the same wallet, and a small drawback is that in order to perform any preparatory steps, the wallets need to be opened and closed before performing the migration, which opens and closes them again.

As for 3., in a sense it provides the worst of all possible worlds, as it allows for passing wallet handles of the same wallet in addition to the drawback of 1., but it allows for the flexibility of doing preparatory steps in between opening the wallets and performing the migration.

Note the limited practicality of preparatory steps mentioned as wallet migration is already performed during the wallet creation (which makes sense, as the schema is inherently coupled to indy-wallet implementation).

Given the above, going with 2. makes the most sense to me, but it may be just because I missed some critical insight.

Reply:
Well, the `Config` and `Credentials` structs are also part of `libvdrtools`, so `WalletHandle` isn't that bad, I guess... 
Nevertheless, after a much closer look I realized that I was very wrong and that the same wallet cannot be opened twice. So I modified the migrator to only take two `WalletHandle`s and error out if they are equal. This also is more in line with what you were saying about the migrator only really dealing with the migration, and not managing wallets.